### PR TITLE
[Core] Support autocompletion on PowerShell

### PR DESCRIPTION
**Pull request by bebound**
**Description**

Resolve https://github.com/Azure/azure-cli/issues/2324
Remake of https://github.com/Azure/azure-cli/pull/13576
Doc pr: https://github.com/MicrosoftDocs/azure-docs-cli/pull/3463

In https://github.com/kislyuk/argcomplete/pull/307, `argcomplete` adds the env `ARGCOMPLETE_USE_TEMPFILES` to redirect its output to file and supports Git Bash. This also make it possible to support PowerShell.

In this PR, I used the same logic which is used in [Git Bash](https://github.com/kislyuk/argcomplete/blob/95cd8bb888840f65a8bda19b4057e7bdd8f02528/argcomplete/bash_completion.d/python-argcomplete#L18-L29): 
- create a temp file, prepare envs
- `argcomplete` writes candidates into the file
- read file and generate `CompletionResult`
- clean envs and temp file

### Envs
-  `ARGCOMPLETE_USE_TEMPFILES = 1`: redirect to file rather than default fd8
- `_ARGCOMPLETE_STDOUT_FILENAME = $completion_file`: output file
- `COMP_LINE = $wordToComplete`: set current full command, ex: `az acc`. It does not include trailing spaces but this won't affect the result. https://github.com/Azure/azure-cli/blob/84e5a177822427461b1a062efed4e4419a00146c/src/azure-cli/az.completion.ps1#L9-L16
- `COMP_POINT = $cursorPosition`: set cursor position, ex: 6
- `_ARGCOMPLETE = 1`: enable `argcomplete`
- `_ARGCOMPLETE_SUPPRESS_SPACE = 0`: make sure when type TAB in `az account`, it becomes `az account `.
  If there is only one suggestion and it is not ends with '=/:"', `argcomplete` adds space at the end when SUPPRESS_SPACE=0.
  In Bash and Fish, it's 1, in Zsh it's 0.

  `argcomplete` is smart. If `nospace` is true, it disables `nospace` temporally and set SUPPRESS_SPACE=1. If `nospace` is false, it set SUPPRESS_SPACE=0 and adds space at the end of suggestion. You can check the [autocomplete file](https://github.com/kislyuk/argcomplete/blob/95cd8bb888840f65a8bda19b4057e7bdd8f02528/argcomplete/shell_integration.py#L8-L54) and [its code](https://github.com/kislyuk/argcomplete/blob/95cd8bb888840f65a8bda19b4057e7bdd8f02528/argcomplete/__init__.py#L622-L629) 

- ``_ARGCOMPLETE_IFS = "`n"``: completion item is split by `IFS`, "`n" is line break in PowerShell, so each line contains one item.

### Document
We should also create a doc for users. The simplest way is to `notepad $PROFILE` and append `Import-Module  "C:\Program Files (x86)\Microsoft SDKs\Azure\CLI2\wbin\az.completion.ps1"`.

[Winget](https://github.com/microsoft/winget-cli/blob/1fbfacc13950de8a17875d40a8beb99fc6ada6c2/doc/Completion.md) and [dotnet](https://learn.microsoft.com/en-us/dotnet/core/tools/enable-tab-autocomplete?WT.mc_id=modinfra-35653-salean#powershell) docs are good example.
> To add tab completion to PowerShell for the .NET CLI, create or edit the profile stored in the variable $PROFILE. For more information, see [How to create your profile](https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_profiles#how-to-create-a-profile) and [Profiles and execution policy](https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_profiles#profiles-and-execution-policy).
> Add the following code to your profile:

To get a bash like completion style, add `Set-PSReadlineKeyHandler -Key Tab -Function MenuComplete` in `$PROFILE`.
> Ctrl+Spacebar         MenuComplete            Complete the input if there is a single completion, otherwise complete the input by selecting from a menu o...     https://learn.microsoft.com/en-us/powershell/module/psreadline/get-psreadlinekeyhandler?view=powershell-7.3

**Testing Guide**
Upgrade `argcomplete`, add `Import-Module  C:\xxx\az.completion.ps1` in `$PROFILE`.
Restart PowerShell, type `az acco`, press tab.

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [ ] The PR title and description has followed the guideline in [Submitting Pull Requests](https://github.com/Azure/azure-cli/tree/dev/doc/authoring_command_modules#submitting-pull-requests).

- [ ] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

- [ ] I adhere to the [Error Handling Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/error_handling_guidelines.md).

### CODE REVIEW

**Comment by yonzhan** [comment]
Support autocompletion on PowerShell

File: build_scripts/windows/scripts/az.completion.psm1 [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+    $env:COMP_LINE = $wordToComplete
+    $env:COMP_POINT = $cursorPosition
+    $env:_ARGCOMPLETE = 1
+    $env:_ARGCOMPLETE_SUPPRESS_SPACE = 0
+    $env:_ARGCOMPLETE_IFS = "`n"
+    az 2>&1 | Out-Null
+    
+    Get-Content $completion_file | ForEach-Object {
+        [System.Management.Automation.CompletionResult]::new($_, $_, 'ParameterValue', $_)
+    }
```
Comment (jiasli) [inline build_scripts/windows/scripts/az.completion.psm1:15]: Could you provide an example of the content of this file?

Comment (bebound) [reply build_scripts/windows/scripts/az.completion.psm1:15]: If the input is `az account `, this file will be:
```
-h
--help
list
set
show
clear
list-locations
get-access-token
lock
management-group
subscription
tenant
alias
accept-ownership-status
```

File: build_scripts/windows/scripts/az.completion.psm1 [hunk 2]
```diff
@@ -0,0 +1,18 @@
+Register-ArgumentCompleter -Native -CommandName az -ScriptBlock {
+    param($commandName, $wordToComplete, $cursorPosition)
```
Comment (bebound) [inline build_scripts/windows/scripts/az.completion.psm1:2]: Dotnet provide an example of how to write Register-ArgumentCompleter : https://learn.microsoft.com/en-us/dotnet/core/tools/enable-tab-autocomplete?WT.mc_id=modinfra-35653-salean#powershell

File: build_scripts/windows/scripts/build.cmd [hunk 3]
```diff
@@ -134,6 +134,7 @@ if %errorlevel% neq 0 goto ERROR
 copy %REPO_ROOT%\build_scripts\windows\resources\CLI_LICENSE.rtf %BUILDING_DIR%
 copy %REPO_ROOT%\build_scripts\windows\resources\ThirdPartyNotices.txt %BUILDING_DIR%
 copy %REPO_ROOT%\NOTICE.txt %BUILDING_DIR%
+copy %REPO_ROOT%\build_scripts\windows\scripts\az.completion.ps1 %BUILDING_DIR%
```
Comment (bebound) [inline build_scripts/windows/scripts/build.cmd:137]: Should we put it to `%BUILDING_DIR%\Scripts\`?
It looks more formal but this folder is created by pip install.

**Pull request closed without merging**